### PR TITLE
feat(ui): render ui to alternative screen

### PR DESCRIPTION
**Pull request by chris-olszewski**
### Description

Changes our UI to render in the alternative screen (think `vim` or `less`)

This provides a few benefits:
 - We can render to the entire terminal without worrying about over-allocating space and removing useful information from the screen.
 - Users won't need to scroll up at the end of a run to see the task logs
 - We no longer use `insert_before` for persisting task logs. This function panics if there isn't an available row to render to (#8072) or rendering terminal logs that have an area that exceeds `u16::MAX` (#7843). Instead our log persisting story is a "simple" `stdout.write_all()`.
 - Removes hacks added to avoid hitting #7843 

We write the logs in a row-wise fasion as `vt100` attempts to optimize performance by using cursor moves to avoid necessary writes. This creates problems when printing multiple terminal screens as the cursor move coordinates will be incorrect.

### Testing Instructions

https://github.com/vercel/turbo/assets/4131117/9746dcd3-188e-48fd-a52d-21d95d0e6e43

Closes TURBO-2979

### CODE REVIEW

**Comment by weyert** [comment]
Would this make it easier to allow selection of log entries?

**Comment by chris-olszewski** [comment]
> Would this make it easier to allow selection of log entries?

Not sure I follow. As in easier to copy/paste from logs from tasks after `turbo` has exited?

**Comment by chris-olszewski** [comment]
> Easier to select from the experimental-ui screen for logging

Ah, no. That broke due to us supporting mouse scroll for scrolling through task outputs. I might need to disable that feature and only allow keybinds. The alternative is to manually implement terminal selection and click support which isn't something I want to tackle at this moment.

**Review by arlyon — APPROVED** [review]

File: crates/turborepo-vt100/src/entire_screen.rs [hunk 1]
```diff
@@ -44,6 +27,53 @@ impl<'a> EntireScreen<'a> {
...
+        let mut wrapping = false;
+        let grid = self.screen.grid();
+        let (rows, _) = self.size();
+        let default = crate::attrs::Attrs::default();
+        grid.all_rows().take(rows).enumerate().map(move |(i, row)| {
+            // number of rows in a grid is stored in a u16 (see Size), so
+            // visible_rows can never return enough rows to overflow here
+            let i = i.try_into().unwrap();
+            let mut contents = vec![];
+            // We don't need final cursor position as long as CRLF is used and not just LF
```
Comment (mehulkar) [inline crates/turborepo-vt100/src/entire_screen.rs:56]: I don't understand this comment. Where would the CRLF or LF be coming from? is it user input? (in which case, is LF not possible on Windows?)

**Review by mehulkar — APPROVED** [review]
The video looks great! Not a blocker, but could we still have colors on the group titles?

CleanShot 2024-05-08 at 11 29 16@2x

File: crates/turborepo-vt100/src/entire_screen.rs (hunk 1 above)
Comment (chris-olszewski) [reply crates/turborepo-vt100/src/entire_screen.rs:56]: When a terminal is in raw mode a LF only will move the cursor down a row, but not return it to the home (zero) column. This leads the string `foo\nbar` being printed in a raw terminal looking like
```
foo
   bar
```
where `foo\r\nbar` will be printed as expected
```
foo
bar
```

The comment is meant to explain why we aren't storing/restoring the final cursor position as we expect the caller to print these lines with a trailing `\r\n`.

**Pull request merged**